### PR TITLE
Feature/basket pagination

### PR DESCRIPTION
**Pull request by MarinaStepanchuk**

### CODE REVIEW

File: src/components/containers/Header/Header.ts [hunk 1]
```diff
@@ -36,7 +36,7 @@ class Header {
                 <span class="header__money__text">${Title.HEADER_TOTAL_MONEY}</span>
                 <span id="headerTotalMoney" class="header__money__value" name="basket-sum">${Symbol.CURRENCY}${this.previousSum}</span>
             </div>
-            <div onclick="window.location.pathname = '/basket'" class="header__basket">
+            <div onclick="window.location.href = '/basket'" class="header__basket">
```
Comment (Ksarelto) [inline src/components/containers/Header/Header.ts:39]: if user click on this button he will reload the page, does it meet requirements of the task?

File: src/components/BasketGoods/BasketGoods.ts [hunk 2]
```diff
@@ -17,6 +18,8 @@ enum Classes {
 class BasketGoods {
   private basket = new Basket();
 
+  private queryParam = new QueryParamsBasket();
```
Comment (Ksarelto) [inline src/components/BasketGoods/BasketGoods.ts:21]: if u use instance of some class in several methods u can create a common value with that instance (but not always ok), but if u use instance of some class in one method better to cteate a variable with instance of necessary class in this method

File: src/components/BasketControl/BasketControl.ts [hunk 3]
```diff
@@ -1,13 +1,87 @@
...
+import QueryParamsBasket from '../../utils/QueryParamsBasket';
+import Basket from '../../utils/Basket';
+import { findElem } from '../../utils/findElem';
+import { DefaultValues } from '../../common.types/enums';
 
 const SUBHEADER_TITLE = 'shopping card'.toUpperCase();
 const AMOUNT_BASKET_GOODS_TITLE = 'items:'.toUpperCase();
 const AMOUNT_BASKET_PAGES_TITLE = 'page:'.toUpperCase();
 
 class BasketHeaderLine {
```
Comment (Ksarelto) [inline src/components/BasketControl/BasketControl.ts:13]: File name is BasketControl but class name is BaketHeaderLine. why?

File: src/components/BasketControl/BasketControl.ts [hunk 4]
```diff
@@ -1,13 +1,87 @@
...
+    }, 0);
+  }
+
+  private getMaxValuePage():number {
+    const totalAmountProducts = this.basket.getBasketList().length;
+    const countProductOnPage = this.getCountProductOnPage();
+    return Math.ceil(totalAmountProducts / countProductOnPage);
+  }
+
+  private getCountProductOnPage() {
```
Comment (Ksarelto) [inline src/components/BasketControl/BasketControl.ts:68]: return types to functions and methods

File: src/components/BasketControl/BasketControl.ts [hunk 5]
```diff
@@ -1,13 +1,87 @@
...
+    return Number(this.queryParam.getParam('limit')) || DefaultValues.PAGINATION_LIMIT;
+  }
+
+  private getCurrentPage() {
+    const maxValuePage = this.getMaxValuePage();
+    const currentPageFromParams = Number(this.queryParam.getParam('page')) || DefaultValues.PAGINATION_PAGE;
+    const currentPage = currentPageFromParams > maxValuePage ? maxValuePage : currentPageFromParams;
+    if (Number(this.queryParam.getParam('page')) > this.getMaxValuePage()) {
```
Comment (Ksarelto) [inline src/components/BasketControl/BasketControl.ts:76]: space around 'if' block

File: src/router/Router.ts [hunk 6]
```diff
@@ -21,8 +21,12 @@ class Router {
     const [,urlPart1, urlPart2] = window.location.pathname.split('/');
     const url = `/${urlPart1 || ''}`;
     const rout = Router.resolveRoute(url);
-    if (RoutesWithRequiredParams.includes(url)) {
-      rout(urlPart2);
+    if (urlPart2) {
+      if (RoutesWithRequiredParams.includes(url)) {
```
[the remark this answers is not in the document]
Comment (MarinaStepanchuk) [reply src/router/Router.ts:25]: Yes, that's clearly better. thank you, I'll fix it.

File: src/components/containers/Header/Header.ts (hunk 1 above)
Comment (MarinaStepanchuk) [reply src/components/containers/Header/Header.ts:39]: I need to cause a page reload to reset the query parameters

File: src/components/BasketGoods/BasketGoods.ts (hunk 2 above)
Comment (MarinaStepanchuk) [reply src/components/BasketGoods/BasketGoods.ts:21]: Okay, I get it.

File: src/components/BasketControl/BasketControl.ts (hunk 3 above)
Comment (MarinaStepanchuk) [reply src/components/BasketControl/BasketControl.ts:13]: inattention)

File: src/components/BasketControl/BasketControl.ts (hunk 4 above)
Comment (MarinaStepanchuk) [reply src/components/BasketControl/BasketControl.ts:68]: I'll add

File: src/utils/QueryParamsBasket.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  public setParam(key: string, value: string): void {
+    this.param.set(key, value);
+  }
+
+  public getParam(key: string): string | null {
+    return this.param.get(key);
+  }
+
+  public sendParams(cb: () => void): void {
+    const baseUrl = `${window.location.protocol}//${window.location.host}${window.location.pathname}`;
```
[the remark this answers is not in the document]
Comment (MarinaStepanchuk) [reply src/utils/QueryParamsBasket.ts]: it does look better)

**Pull request merged**